telemetry/telemetry.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) 2024 The VeChainThor developers
+
+// Distributed under the GNU Lesser General Public License v3.0 software license, see the accompanying
+// file LICENSE or <https://www.gnu.org/licenses/lgpl-3.0.html>
+
+package telemetry
+
+import "net/http"
+
+// telemetry is a singleton service that provides global access to a set of meters
+// it wraps multiple implementations and defaults to a no-op implementation
+var telemetry = defaultNoopTelemetry() // defaults to a Noop implementation of the telemetry service
+
+// Telemetry defines the interface for telemetry service implementations
+type Telemetry interface {
+	GetOrCreateCountMeter(name string) CountMeter
+	GetOrCreateCountVecMeter(name string, labels []string) CountVecMeter
+	GetOrCreateGaugeMeter(name string) GaugeMeter
+	GetOrCreateGaugeVecMeter(name string, labels []string) GaugeVecMeter
+	GetOrCreateHistogramMeter(name string, buckets []int64) HistogramMeter
+	GetOrCreateHistogramVecMeter(name string, labels []string, buckets []int64) HistogramVecMeter
+	GetOrCreateHandler() http.Handler
+}
+
+// HTTPHandler returns the http handler for retrieving metrics
+func HTTPHandler() http.Handler {
+	return telemetry.GetOrCreateHandler()
+}
+
+// Define standard buckets for histograms
+var (
+	Bucket10s      = []int64{0, 500, 1000, 2000, 3000, 4000, 5000, 7500, 10_000}
+	BucketHTTPReqs = []int64{0, 150, 300, 450, 600, 900, 1200, 1500, 3000}
+)
+
+// HistogramMeter represents the type of metric that is calculated by aggregating
+// as a Histogram of all reported measurements over a time interval.
+type HistogramMeter interface {
+	Observe(int64)
+}
+
+func Histogram(name string, buckets []int64) HistogramMeter {
+	return telemetry.GetOrCreateHistogramMeter(name, buckets)
+}
+
+// HistogramVecMeter same as the Histogram but with labels
+type HistogramVecMeter interface {
+	ObserveWithLabels(int64, map[string]string)
+}
+
+func HistogramVec(name string, labels []string, buckets []int64) HistogramVecMeter {
+	return telemetry.GetOrCreateHistogramVecMeter(name, labels, buckets)
+}
+
+// CountMeter is a cumulative metric that represents a single monotonically increasing counter
+// whose value can only increase or be reset to zero on restart.
+type CountMeter interface {
```

Comment:
How about keeping it the same with `prometheus.Gauge`?

Reply:
It's a nice simplification and I think it can be done, but counter and gauges can be represented in different ways in grafana. Would avoid it. My concern is that doing simplification for a potential gain might bite us in the arse further down the line.

Reply:
I think the confusing part is that our `Gauge(int)` method is actually an `Add(int)` just like Counters.
I'll change the gauge struct to it's 2 core usages,`Set` and `Add` there's is also `Sub` and `Inc` but I think it would be pref to add those as needed ?

Reply:
TLDR; We need Gauge to do Add(-1) which Counters don't do.